tenable/ot/graphql/iterators.py

```diff
@@ -1,51 +1,78 @@
-'''
+"""
 GraphQL Tenable.ot API iterator.
-'''
+"""
+import json
+import typing
+from typing import Type
+
 from marshmallow.utils import EXCLUDE
 from restfly.iterator import APIIterator
+
 from tenable.ot.graphql.definitions import (
     GraphqlErrorSchema,
-    GraphqlParsingError
+    GraphqlParsingError,
+    GraphObject,
+    NodesSchema,
 )
 
+if typing.TYPE_CHECKING:
+    from tenable.ot.session import TenableOT
+
 
 class OTGraphIterator(APIIterator):
-    '''
-    Iterator class over Tenable.ot GraphQL connetions.
-    '''
+    """
+    Iterator class over Tenable.ot GraphQL connections.
+    """
 
-    def __init__(self, api, graph_object, **kwargs):
+    def __init__(
+        self,
+        api: "TenableOT",
+        graph_object: GraphObject,
+        **kwargs,
+    ):
         self._graph_object = graph_object
+        self.api = api
         super().__init__(api, **kwargs)
 
-    def _get_page(self):
-        '''
+    def _get_page(self) -> Type[NodesSchema]:
+        """
         Retrieves the next page of data.
-        '''
+        """
         graph_full_object = {
-            'query': self._graph_object.query,
-            'variables': self._graph_object.query_variables
+            "query": self._graph_object.query,
+            "variables": json.dumps(self._graph_object.query_variables),
         }
+        graphql_response_object = self._perform_request(graph_full_object)
 
-        resp = self._api.graphql(**graph_full_object)
-        if 'error' in resp:
+        self.page = self._graph_object.resp_schema().load(
+            graphql_response_object, unknown=EXCLUDE
+        )
+        if graphql_response_object.get("pageInfo"):
+            self._graph_object.query_variables["startAt"] = graphql_response_object[
+                "pageInfo"
+            ].get("endCursor", None)
+
+        return self.page
+
+    def _perform_request(
+        self,
+        graph_full_object: dict,
+    ) -> dict:
+        resp = self.api.graphql(**graph_full_object)
```

Comment:
I think you should add timeout for request not responding (Ip not reachable and so on...)
Not sure if it should be configurable but maybe 10 sec? or maybe 20 sec?

Reply:
If a timeout is desired, then one should/count be set in the TenableOT class (which can be overloaded at instantiation) or can be passed to the .graphql method directly (using the timeout param)

https://restfly.readthedocs.io/en/latest/api/session.html#restfly.session.APISession._timeout

https://requests.readthedocs.io/en/latest/user/advanced/?highlight=timeout#timeouts